sdk/search/Azure.Search.Documents/tests/Samples/FieldBuilder/FieldBuilder.cs

```diff
@@ -0,0 +1,451 @@
+                }
+
+                IDataTypeInfo dataTypeInfo = GetDataTypeInfo(prop.PropertyType, namingPolicy);
+
+                return dataTypeInfo.Match(
+                    onUnknownDataType: () => throw FailOnUnknownDataType(),
+                    onSimpleDataType: CreateSimpleField,
+                    onComplexDataType: CreateComplexField);
+            }
+
+            return info.Properties.Select(BuildField).Where(field => field != null).ToArray();
+        }
+
+        private static IDataTypeInfo GetDataTypeInfo(Type propertyType, JsonNamingPolicy namingPolicy)
+        {
+            static bool IsNullableType(Type type) =>
+                type.IsConstructedGenericType && type.GetGenericTypeDefinition() == typeof(Nullable<>);
+
+            if (s_primitiveTypeMap.TryGetValue(propertyType, out SearchFieldDataType SearchFieldDataType))
+            {
+                return DataTypeInfo.Simple(SearchFieldDataType);
+            }
+            else if (IsNullableType(propertyType))
+            {
+                return GetDataTypeInfo(propertyType.GenericTypeArguments[0], namingPolicy);
+            }
+            else if (TryGetEnumerableElementType(propertyType, out Type elementType))
+            {
+                IDataTypeInfo elementTypeInfo = GetDataTypeInfo(elementType, namingPolicy);
+                return DataTypeInfo.AsCollection(elementTypeInfo);
+            }
+            else if (ObjectInfo.TryGet(propertyType, out ObjectInfo info))
+            {
+                return DataTypeInfo.Complex(SearchFieldDataType.Complex, propertyType, info);
+            }
+            else
+            {
+                return DataTypeInfo.Unknown;
+            }
+        }
+
+        private static bool TryGetEnumerableElementType(Type candidateType, out Type elementType)
+        {
+            static Type GetElementTypeIfIEnumerable(Type t) =>
+                t.IsConstructedGenericType && t.GetGenericTypeDefinition() == typeof(IEnumerable<>)
+                    ? t.GenericTypeArguments[0]
+                    : null;
+
+            elementType = GetElementTypeIfIEnumerable(candidateType);
+            if (elementType != null)
+            {
+                return true;
+            }
+            else
+            {
+                TypeInfo ti = candidateType.GetTypeInfo();
+                var listElementTypes = ti
+                    .ImplementedInterfaces
+                    .Select(GetElementTypeIfIEnumerable)
+                    .Where(p => p != null)
+                    .ToList();
+
+                if (listElementTypes.Count == 1)
+                {
+                    elementType = listElementTypes[0];
+                    return true;
+                }
+                else
+                {
+                    return false;
+                }
+            }
+        }
+
+        private interface IDataTypeInfo
+        {
+            T Match<T>(
```

Comment:
This is more of a `Dispatch` than `Match`, right?

Reply:
Straight port of track 1. Track 2 will work much differently.